async-raft/tests/add_remove_voter.rs

```diff
@@ -0,0 +1,143 @@
+    let left_members = hashset![0, 1, 2, 3];
+
+    // Setup test dependencies.
+    let config = Arc::new(Config::build("test".into()).validate().expect("failed to build Raft config"));
+    let router = Arc::new(RaftRouter::new(config.clone()));
+    router.new_raft_node(0).await;
+
+    // Assert all nodes are in non-voter state & have no entries.
+    let mut want = 0;
+    router
+        .wait_for_metrics(&0u64, |x| x.last_log_index == want, timeout, &format!("n{}.last_log_index -> {}", 0, 0))
+        .await?;
+    router
+        .wait_for_metrics(
+            &0u64,
+            |x| x.state == State::NonVoter,
+            timeout,
+            &format!("n{}.state -> {:?}", 4, State::NonVoter),
+        )
+        .await?;
+
+    router.assert_pristine_cluster().await;
+
+    // Initialize the cluster, then assert that a stable cluster was formed & held.
+    tracing::info!("--- initializing cluster");
+    router.initialize_from_single_node(0).await?;
+    want = 1;
+
+    wait_log(router.clone(), &hashset![0], want).await?;
+    router.assert_stable_cluster(Some(1), Some(want)).await;
+
+    // Sync some new nodes.
+    router.new_raft_node(1).await;
+    router.new_raft_node(2).await;
+    router.new_raft_node(3).await;
+    router.new_raft_node(4).await;
+
+    tracing::info!("--- adding new nodes to cluster");
+    let mut new_nodes = futures::stream::FuturesUnordered::new();
+    new_nodes.push(router.add_non_voter(0, 1));
+    new_nodes.push(router.add_non_voter(0, 2));
+    new_nodes.push(router.add_non_voter(0, 3));
+    new_nodes.push(router.add_non_voter(0, 4));
+    while let Some(inner) = new_nodes.next().await {
+        inner?;
+    }
+
+    wait_log(router.clone(), &all_members, want).await?;
+
+    tracing::info!("--- changing cluster config");
+    router.change_membership(0, all_members.clone()).await?;
+    want += 2; // 2 member-change logs
+
+    wait_log(router.clone(), &all_members, want).await?;
+    router.assert_stable_cluster(Some(1), Some(want)).await; // Still in term 1, so leader is still node 0.
+
+    // Send some requests
+    router.client_request_many(0, "client", 100).await;
+    want += 100;
+
+    wait_log(router.clone(), &all_members, want).await?;
+
+    // Remove Node 4
+    tracing::info!("--- remove n{}", 4);
+    router.change_membership(0, left_members.clone()).await?;
+    want += 2; // two member-change logs
+
+    wait_log(router.clone(), &left_members, want).await?;
+    router
+        .wait_for_metrics(
+            &4u64,
+            |x| x.state == State::NonVoter,
+            timeout,
+            &format!("n{}.state -> {:?}", 4, State::NonVoter),
+        )
+        .await?;
+
+    // Send some requests
+    router.client_request_many(0, "client", 100).await;
+    want += 100;
+
+    wait_log(router.clone(), &left_members, want).await?;
+
+    // log will not be sync to removed node
+    let x = router.latest_metrics().await;
+    assert!(x[4].last_log_index < want);
```

Comment:
Feels hard to be accurate about this check:

Since the replication is in another tokio task, I am not sure whether we have a way to stop the replication at once when the `uniform` membership is committed.

The entries are sent to `replication_buffer`. And the replication task(thread) drains `replication_buffer` and sends them out.

At the receiving end, it is possible to stop receiving further log entries when it is aware that it has been removed from memberships.
But it somehow violates the assumption about Follower/NonVoter: no matter it is a follower or non-voter it always accepts logs sent from some leader.